### ISSUE TITLE
SPARQL Update: DELETE, error: "Message: CL...: The array extend 1048576 is less than requested 1310721"

### ISSUE THREAD

**Issue by fgiasson** [openlink/virtuoso-opensource#409]
I am getting that error:

```
Message: CL...: The array extend 1048576 is less than requested 1310721
```

When sending that SPARUL query:

```
SPARQL 
DELETE FROM <http://freebase.com> 
{   
  ?s <http://rdf.freebase.com/ns/type.object.type> ?o .  
}
WHERE
{  
  graph <http://freebase.com> {    
    ?s <http://rdf.freebase.com/ns/type.object.type> ?o .  
  }
};
```

Before getting that error, I was getting that other error:

```
FRVEC: array in for vectored over max vector length 2000000 > 1000000
```

However, increasing the `MaxVectorSize` setting to 3M solved that issue.

@HughWilliams you said that the `CL...` error was due related to a clustering issue in one mailthread a read. However, I am clearly using the Open Source version:

```
root@SDM:~# virtuoso-t -?
Virtuoso Open Source Edition (Column Store) (multi threaded)
Version 7.2.0-rc1.3211-pthreads as of Jan 28 2015
Compiled for Linux (x86_64-pc-linux-gnu)
Copyright (C) 1998-2015 OpenLink Software
```

**Comment by kidehen** [comment]
Fred,

Call vectorization scoped to threads is a feature of the column-store, 
which is open source.

## 

Regards,

Kingsley Idehen 
Founder & CEO
OpenLink Software
Company Web: http://www.openlinksw.com
Personal Weblog 1: http://kidehen.blogspot.com
Personal Weblog 2: http://www.openlinksw.com/blog/~kidehen
Twitter Profile: https://twitter.com/kidehen
Google+ Profile: https://plus.google.com/+KingsleyIdehen/about
LinkedIn Profile: http://www.linkedin.com/in/kidehen
Personal WebID: http://kingsley.idehen.net/dataspace/person/kidehen#this

**Comment by fgiasson** [comment]
@kidehen, good thanks. In any case, do you have an idea why I am getting that error?

**Comment by kidehen** [comment]
You exceeded the threshold of the default (or existing) setting. 
Freebase dataset contains a lot of data etc..

**Comment by fgiasson** [comment]
@kidehen is there are indication of which setting is related to that error?

**Comment by kidehen** [comment]
You indicated "increasing the |MaxVectorSize| setting to 3M solved that 
issue".  Meaning you increased the MaxVectorSize to 3M.

[1] http://docs.openlinksw.com/virtuoso/databaseadmsrv.html#confvectexec 
-- about vectorized execution related settings.

**Comment by HughWilliams** [comment]
@fgiasson:  The last time I saw the "CL...: The array extend 1048576 is less than requested 1310721" error the user was performing delete operations and got around it by splitting the delete operations with "limit" clause to avoid hitting the limit.

Do you have any idea as to the count of triples being deleted ? Also, if these deletes are being performed against freebase, we could try setting up the same test locally, thus which freebase dataset(s) are you using ?

**Comment by fgiasson** [comment]
HI @HughWilliams,

Do you know the limit he used?

This is a subset of freebase that is in that dataset, but the number of triples that should be removed with that query is `255 011 937`.

@kidehen the MaxVectorSize setting solved the `FRVEC: array in for vectored over max vector length 2000000 > 1000000` error, not that one, sorry for the confusion and if the question was not clear.

Could there be a setting that could affect that second error?

**Comment by fgiasson** [comment]
@HughWilliams the freebase dump I am using is Google's:

```
https://developers.google.com/freebase/data
```

**Comment by HughWilliams** [comment]
@fgiasson: Not sure into what size the user split there deletes into, but would suggest splitting it in 2 to start with, see if it is successful, and if not split the split files in 2 again etc until it is successful ...

It seems we already have the freebase datasets download on a dev server so will check if it has already been loaded anywhere were I can test the delete ...

**Comment by fgiasson** [comment]
@HughWilliams yeah, it is about 1M, after that it returns the error. Since I have 255M triples to download, I will wait for a better solution.

So you know, the server I am working with has 128G Ram and 8 cores. So it should be able to handle this I guess (at least, it loads it).

**Comment by randykerber** [comment]
Is there an answer to this?

I am getting this message:
`Virtuoso 42000 Error CL...: The array extend 1048576 is less than requested 1310721`

The original poster made it clear that this was not because of `MaxVectorSize`.

What option needs to be adjusted to permit more triples to be deleted at one time?

**Comment by HughWilliams** [comment]
There is no immediate solution to this problem of deleting large graphs/triples ...

As said previously, you could try splitting the delete operations in smaller sizes/batches ...

Depending on how much free memory is available on the server, you could also try increasing the `MaxMemPoolSize` param in the INI file as detailed [in the manual](http://docs.openlinksw.com/virtuoso/dbadm/).

**Comment by asanchez75** [comment]
I had the same problem but it is possible to workaround it by running SQL queries directly against the underlying database. It would be something like:

`DELETE FROM DB.DBA.RDF_QUAD where P = iri_to_id('http://rdf.freebase.com/ns/type.object.type') and G = iri_to_id('http://freebase.com')`

**Comment by rapw3k** [comment]
Hi, 
I had exactly the same error when trying to update large number of triples with 

`WITH <http://w3id.org/...> DELETE { ?s <http://w3id.org/... ?o } INSERT { ?s <http://w3id.org/... ?o2 } WHERE { ...}`

I had to use the workaround proposed by @asanchez75 working directly on the underlying database, and it worked...

In any case, its really missing clear information about how to overcome such errors, i.e., what configuration parameters should be changed, what are the exact limitation in virtuoso, etc.

**Comment by HughWilliams** [comment]
See the [document on deletion of large graphs in Virtuoso](http://vos.openlinksw.com/owiki/wiki/VOS/VirtTipsAndTricksGuideDeleteLargeGraphs) ...
